<a href="https://colab.research.google.com/github/yhy0519/Machine_learning_of_Python/blob/main/python_m05_03_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **■ 신경망(Neural Network)**

          사람의 뇌   ------------------------>   컴퓨터를 이용한 지능처리
              ↓                                             ↓   
       생물학적 신경망내에서                   인공신경망에서는 가중치라는 것으로
    반복적인 시그널이 발생할 때              기억의 효과를 대체할 수 있음을 설명했다.
    그 시그널을 기억하는 일종의
         학습효과가 있다

※ 파라미터와 하이퍼 파라미터의 차이 ? (수제비3-12)

 ①. 파라미터 ? 
 
    모델 내부에서 확인이 가능한 변수로 데이터를 통해서 산출이 가능한 값
    예측을 수행할 때, 모델에 의해 요귀되어지는 값들
    사람의 수작업으로 측정되지 않음
	
	예: 신경망에서의 가중치, 회귀분석에서의 결정계수, 서포트 벡터 머신에서의 서포트 벡터

 ②. 하이퍼 파라미터 ? 

    모델에서 외적인 요소로 데이터 분석을 통해 얻어지는 값이 아니라 사용자가 직접 설정해주는 값
    모델의 파라미터값을 측정하기 위해 알고리즘 구현 과정에서 사용
    하이퍼 파라미터는 주로 알고리즘 사용자에 의해 결정
				
	예: 신경망에서의 학습률(러닝레이트), 
	     knn 에서의 k 값, 
	     의사결정트리에서의 나무의 깊이,
	     서포트 벡터머신에서의 C 값과 gamma 값
	

 최적의 하이퍼 파라미터를 알아내는 기법 ? 

	파이썬  →  grid search 기법
	R  →  caret 패키지

** 퍼셉트론 ? 뇌의 신경세포 하나를 컴퓨터로 흉내낸 것

		            		반복 학습 (중요도)
				              	↓
    1. 뇌 : 신호  ------------------------------------->  암기

					                  가중치
				                  	  ↓             1 신호가 흐른다.
    2. 인공신경망 : 신호  --------------------------->  
								                       0 신호가 흐르지 않는다.

              컴퓨터                                      뇌
                ↓                                        ↓
    정보를 메모리의 특정 위치에 저장      뇌는 정보를 저장하는 공간이 따로 없다.
		                            					  ↓
                                         신경세포(뉴런)의 연결관계를 변경하는 방식으로 정보를 저장한다.

    - R : nearnlnet 함수 구현 : 콘크리트의 강도를 예측
    - 파이썬 : 사이킷런의 neural_network 함수 구현 : 콘크리트 강도를 예측
		 1. 예측 : 콘크리트 데이터
		 2. 분류 : 타이타닉 데이터

문제39. 콘크리트 데이터를 파이썬의 신경망으로 강도를 예측을 하시오 !

    ** 콘크리트 데이터 설명
    1. cement: 콘크리트의 총량
    2. slag  :  시멘트 
    3. ash   :  분 (시멘트)
    4. water :  물
    5. superplasticizer :  고성능 감수재(콘크리트 강도를 높이는 첨가제)
    6. coarse aggregate :  굵은 자갈
    7. fine  aggregate :  잔 자갈
    8. aging time  :  숙성시간 
    9. strength : 콘크리트 강도

In [1]:
### 기본 라이브러리 불러오기
import pandas as pd

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = pd.read_csv('/content/drive/MyDrive/data/concrete.csv')
df

cement   slag    ash  water  ...  coarseagg  fineagg  age  strength
0      141.3  212.0    0.0  203.5  ...      971.8    748.5   28     29.89
1      168.9   42.2  124.3  158.3  ...     1080.8    796.2   14     23.51
2      250.0    0.0   95.7  187.4  ...      956.9    861.2   28     29.22
3      266.0  114.0    0.0  228.0  ...      932.0    670.0   28     45.85
4      154.8  183.4    0.0  193.3  ...     1047.4    696.7   28     18.29
...      ...    ...    ...    ...  ...        ...      ...  ...       ...
1025   135.0    0.0  166.0  180.0  ...      961.0    805.0   28     13.29
1026   531.3    0.0    0.0  141.8  ...      852.1    893.7    3     41.30
1027   276.4  116.0   90.3  179.6  ...      870.1    768.3   28     44.28
1028   342.0   38.0    0.0  228.0  ...      932.0    670.0  270     55.06
1029   540.0    0.0    0.0  173.0  ...     1125.0    613.0    7     52.61

[1030 rows x 9 columns]

In [2]:
'''
[Step 2] 데이터 탐색/ 전처리
'''
# 결측치 확인
df.isnull().sum()

cement          0
slag            0
ash             0
water           0
superplastic    0
coarseagg       0
fineagg         0
age             0
strength        0
dtype: int64

In [3]:
df.head()

cement   slag    ash  water  superplastic  coarseagg  fineagg  age  strength
0   141.3  212.0    0.0  203.5           0.0      971.8    748.5   28     29.89
1   168.9   42.2  124.3  158.3          10.8     1080.8    796.2   14     23.51
2   250.0    0.0   95.7  187.4           5.5      956.9    861.2   28     29.22
3   266.0  114.0    0.0  228.0           0.0      932.0    670.0   28     45.85
4   154.8  183.4    0.0  193.3           9.1     1047.4    696.7   28     18.29

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cement        1030 non-null   float64
 1   slag          1030 non-null   float64
 2   ash           1030 non-null   float64
 3   water         1030 non-null   float64
 4   superplastic  1030 non-null   float64
 5   coarseagg     1030 non-null   float64
 6   fineagg       1030 non-null   float64
 7   age           1030 non-null   int64  
 8   strength      1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


In [5]:
'''
[Step 3] 분석에 사용할 속성을 선택
'''
# 속성(변수) 선택
X=df[['cement', 'slag','ash','water','superplastic','coarseagg','fineagg','age']]  # 독립 변수 X

y=df['strength']           # 종속 변수 Y

In [6]:
'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''
X.head()

cement   slag    ash  water  superplastic  coarseagg  fineagg  age
0   141.3  212.0    0.0  203.5           0.0      971.8    748.5   28
1   168.9   42.2  124.3  158.3          10.8     1080.8    796.2   14
2   250.0    0.0   95.7  187.4           5.5      956.9    861.2   28
3   266.0  114.0    0.0  228.0           0.0      932.0    670.0   28
4   154.8  183.4    0.0  193.3           9.1     1047.4    696.7   28

In [7]:
# 설명 변수 데이터를 표준화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
print(X)
print('\n')
print(X.shape)
print('\n')
print(type(X))

[[-1.33901711  1.60144087 -0.84714393 ... -0.01439767 -0.31296991
  -0.27973311]
 [-1.07479007 -0.36754132  1.09607803 ...  1.38814131  0.28226038
  -0.50146528]
 [-0.29838379 -0.85688789  0.64896501 ... -0.20612089  1.09337085
  -0.27973311]
 ...
 [-0.04564488  0.4882354   0.56454507 ... -1.32300514 -0.06589318
  -0.27973311]
 [ 0.58237302 -0.41624406 -0.84714393 ... -0.52651741 -1.29254178
   3.55306569]
 [ 2.47791487 -0.85688789 -0.84714393 ...  1.9568773  -2.00382326
  -0.61233136]]


(1030, 8)


<class 'numpy.ndarray'>


In [8]:
# train data 와 test data로 구분(8:2 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

train data 개수:  (824, 8)
test data 개수:  (206, 8)




In [9]:
'''
[Step 5] 신경망 모형 - sklearn 사용
'''
##### from sklearn.neural_network
##### 1. 분류 : MLPClassifier
##### 2. 수치예측 : MLPRegressor

from sklearn.neural_network import MLPRegressor  # 수치예측

# neural network 모델 적합
clf = MLPRegressor(random_state=0, max_iter=500)  # 학습을 500번 하겠다.

In [10]:
'''
[Step 6] train data를 가지고 모형 학습
'''
# train data를 가지고 모형 학습
clf.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(100,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=500,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=0, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [11]:
'''
[Step 7] test data를 가지고 y_hat을 예측
'''
# test data를 가지고 y_hat을 예측 (분류) 
y_hat = clf.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

[36.47812292 11.06762334 45.98514134 23.52352179 50.73694178 53.30473313
 17.95362979 42.53714652 23.10688494 50.79253454]
[47.4  10.35 49.99 14.7  47.82 51.06 13.29 52.82 22.5  47.74]




In [12]:
'''
[Step 8] 모형 성능 평가 - 상관관계 계산
'''
import scipy.stats as stats
svm_report =  stats.pearsonr(y_test, y_hat)  # 상관계수 확인하는 코드
print(svm_report)   # 상관계수, p-value(H0 : 상관계수가 0 이다. vs H1 : 상관계수가 0이 아니다.)

(0.9042961287774729, 2.385285961523617e-77)


문제40. 위의 신경망의 성능을 더 올리시오 !  
	   성능을 더 올리라는 것은 뉴런의 개수를 늘리거나 층을 더 늘리거나 하는
	   신경망에 연관된 하이퍼 파라미터를 알아내야 합니다.  
	   하이퍼 파라미터를 자동으로 알아내게 하는 파이썬의 기능인 grid search 를 이용해서 알아내시오 !

In [ ]:
# 힌트코드:

from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import GridSearchCV  # grid search 기능을 사용하기 위해서

param_grid = [
        {
            'activation' : ['identity', 'logistic', 'tanh', 'relu'],   # 활성화 함수의 종류 4가지
            'solver' : ['lbfgs', 'sgd', 'adam'],   # 옵티마이져(최적화)의 종류 3가지
            'hidden_layer_sizes': [
             (1,),(2,),(3,),(4,),(5,),(6,),(7,),(8,),(9,),(10,),(11,), (12,),(13,),(14,),(15,),(16,),(17,),(18,),(19,),(20,),(21,)
             ]
        }
       ]

# 옵티마이져 ? global minima 를 찾아가기 위한 방법론(수학식)을 코드로 구현한 것
# lbfgs 옵티마이져 ? 가장 인기있는 옵티마이져이고 computer memory 의 제한된 양만을
#                          사용해서 global minima 를 찾아가도록 코드가 구현되어 있다.
# hidden_layer_size = (뉴런수, 층수) , 위의 예제에서는 뉴런수만 정해줬다. 층수는 X

# 모델을 생성한다.
clf = GridSearchCV(MLPRegressor(), param_grid, cv=3, n_jobs = -1, verbose = 2 )
# 설명: clf = GridSearchCV(모델명, grid 파라미터 값, k-holdout 개수 설정, 
#                                  n_jobs = -1, verbose = 2 )
# cv=3 은 k-holdout 의 k 값을 3으로 해서 훈련 데이터를 3개의 파티션으로 나누고
# 2개를 훈련 데이터로 사용하고 나머지 한개를 테스트 데이터로 사용하겠다는 것
# n_jobs = -1 를 사용하게 되면 최적의 하이퍼 파라미터를 알아내는 작업 전체를 보여주는게
# 아니라 요약해서 보여준다.
# verbose = 2 는 gridSearch 하는 과정을 보여준다.

clf.fit(X_train, y_train)  # 위에서 만든 모델을 훈련시킨다.


print("Best parameters set found on development set:") 
print(clf.best_params_)  # 최고의 하이퍼 파라미터 조합을 출력해준다.

In [14]:
### 기본 라이브러리 불러오기
import pandas as pd

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = pd.read_csv('/content/drive/MyDrive/data/concrete.csv')
df


cement   slag    ash  water  ...  coarseagg  fineagg  age  strength
0      141.3  212.0    0.0  203.5  ...      971.8    748.5   28     29.89
1      168.9   42.2  124.3  158.3  ...     1080.8    796.2   14     23.51
2      250.0    0.0   95.7  187.4  ...      956.9    861.2   28     29.22
3      266.0  114.0    0.0  228.0  ...      932.0    670.0   28     45.85
4      154.8  183.4    0.0  193.3  ...     1047.4    696.7   28     18.29
...      ...    ...    ...    ...  ...        ...      ...  ...       ...
1025   135.0    0.0  166.0  180.0  ...      961.0    805.0   28     13.29
1026   531.3    0.0    0.0  141.8  ...      852.1    893.7    3     41.30
1027   276.4  116.0   90.3  179.6  ...      870.1    768.3   28     44.28
1028   342.0   38.0    0.0  228.0  ...      932.0    670.0  270     55.06
1029   540.0    0.0    0.0  173.0  ...     1125.0    613.0    7     52.61

[1030 rows x 9 columns]

In [15]:

'''
[Step 2] 데이터 탐색/ 전처리
'''
# 결측치 확인
df.isnull().sum()

cement          0
slag            0
ash             0
water           0
superplastic    0
coarseagg       0
fineagg         0
age             0
strength        0
dtype: int64

In [16]:
df.head()

cement   slag    ash  water  superplastic  coarseagg  fineagg  age  strength
0   141.3  212.0    0.0  203.5           0.0      971.8    748.5   28     29.89
1   168.9   42.2  124.3  158.3          10.8     1080.8    796.2   14     23.51
2   250.0    0.0   95.7  187.4           5.5      956.9    861.2   28     29.22
3   266.0  114.0    0.0  228.0           0.0      932.0    670.0   28     45.85
4   154.8  183.4    0.0  193.3           9.1     1047.4    696.7   28     18.29

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cement        1030 non-null   float64
 1   slag          1030 non-null   float64
 2   ash           1030 non-null   float64
 3   water         1030 non-null   float64
 4   superplastic  1030 non-null   float64
 5   coarseagg     1030 non-null   float64
 6   fineagg       1030 non-null   float64
 7   age           1030 non-null   int64  
 8   strength      1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


In [18]:
'''
[Step 3] 분석에 사용할 속성을 선택
'''
# 속성(변수) 선택
X=df[['cement', 'slag','ash','water','superplastic','coarseagg','fineagg','age']]  #독립 변수 X

y=df['strength']                      #종속 변수 Y

In [20]:
'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''
# 설명 변수 데이터를 표준화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
print(X)
print('\n')
print(type(X))
print('\n')

[[-1.33901711  1.60144087 -0.84714393 ... -0.01439767 -0.31296991
  -0.27973311]
 [-1.07479007 -0.36754132  1.09607803 ...  1.38814131  0.28226038
  -0.50146528]
 [-0.29838379 -0.85688789  0.64896501 ... -0.20612089  1.09337085
  -0.27973311]
 ...
 [-0.04564488  0.4882354   0.56454507 ... -1.32300514 -0.06589318
  -0.27973311]
 [ 0.58237302 -0.41624406 -0.84714393 ... -0.52651741 -1.29254178
   3.55306569]
 [ 2.47791487 -0.85688789 -0.84714393 ...  1.9568773  -2.00382326
  -0.61233136]]


<class 'numpy.ndarray'>




In [21]:
# train data 와 test data로 구분(8:2 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

train data 개수:  (824, 8)
test data 개수:  (206, 8)




In [22]:
'''
[Step 5] 신경망 모형 - sklearn 사용
'''
##### from sklearn.neural_network
##### 1. 분류 : MLPClassifier
##### 2. 수치예측 : MLPRegressor

from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import GridSearchCV  # grid search 기능을 사용하기 위해서

param_grid = [
        {
            'activation' : ['identity', 'logistic', 'tanh', 'relu'],   # 활성화 함수의 종류 4가지
            'solver' : ['lbfgs', 'sgd', 'adam'],   # 옵티마이져(최적화)의 종류 3가지
            'hidden_layer_sizes': [
             (1,),(2,),(3,),(4,),(5,),(6,),(7,),(8,),(9,),(10,),(11,), (12,),(13,),(14,),(15,),(16,),(17,),(18,),(19,),(20,),(21,)
             ]
        }
       ]


clf = GridSearchCV(MLPRegressor(), param_grid, cv=3, n_jobs = -1, verbose = 2 )

In [27]:
'''
[Step 6] train data를 가지고 모형 학습
'''
# train data를 가지고 모형 학습
clf.fit(X_train, y_train)

Fitting 3 folds for each of 252 candidates, totalling 756 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  70 tasks      | elapsed:    6.0s
[Parallel(n_jobs=-1)]: Done 312 tasks      | elapsed:   36.9s
[Parallel(n_jobs=-1)]: Done 718 tasks      | elapsed:  1.8min
[Parallel(n_jobs=-1)]: Done 756 out of 756 | elapsed:  2.0min finished
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:470: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


GridSearchCV(cv=3, error_score=nan,
             estimator=MLPRegressor(activation='relu', alpha=0.0001,
                                    batch_size='auto', beta_1=0.9, beta_2=0.999,
                                    early_stopping=False, epsilon=1e-08,
                                    hidden_layer_sizes=(100,),
                                    learning_rate='constant',
                                    learning_rate_init=0.001, max_fun=15000,
                                    max_iter=200, momentum=0.9,
                                    n_iter_no_change=10,
                                    nesterovs_momentum=True, power_t=0.5,
                                    random_state=...
                                    warm_start=False),
             iid='deprecated', n_jobs=-1,
             param_grid=[{'activation': ['identity', 'logistic', 'tanh',
                                         'relu'],
                          'hidden_layer_sizes': [(1,), (2,), (3,), (4,)

In [28]:
print("Best parameters set found on development set:") 
print(clf.best_params_)  # 최고의 하이퍼 파라미터 조합을 출력해준다.

Best parameters set found on development set:
{'activation': 'tanh', 'hidden_layer_sizes': (19,), 'solver': 'lbfgs'}


In [29]:
'''
[Step 7] test data를 가지고 y_hat을 예측
'''
# test data를 가지고 y_hat을 예측 (분류) 
y_hat = clf.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

[41.37003251  8.36959465 56.45780139 14.94350144 45.97480399 46.942434
 10.6666804  47.71676041 23.02749951 46.10645193]
[47.4  10.35 49.99 14.7  47.82 51.06 13.29 52.82 22.5  47.74]




In [30]:
'''
[Step 8] 모형 성능 평가 - 상관관계 계산
'''
import scipy.stats as stats
svm_report =  stats.pearsonr(y_test, y_hat)  # 상관계수 확인하는 코드
print(svm_report)   # 상관계수, p-value

(0.9610749947497635, 6.329878557246171e-116)


Step 5 에서 clf = GridSearchCV(MLPRegressor(random_state=10), param_grid, cv=3, n_jobs = -1, verbose = 2 ] 로 random_state 값을 설정해서 실행하면 값 동일하게 출력.

MLPRegressor(random_state=10) 인 경우 

    Best parameters set found on development set: {'activation': 'tanh', 'hidden_layer_sizes': (21,), 'solver': 'lbfgs'}
    상관계수, p-value: (0.964207283401068, 1.4270030829531285e-119)

문제41. 모델 생성할때 설정하는 random_stats 도 gridsearch 해서 알아내게 하려면 ?

In [32]:
### 기본 라이브러리 불러오기
import pandas as pd

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = pd.read_csv('/content/drive/MyDrive/data/concrete.csv')

'''
[Step 2] 데이터 탐색/ 전처리
'''
# 결측치 확인
#df.isnull().sum()
#df.head()
#df.info()

'''
[Step 3] 분석에 사용할 속성을 선택
'''
# 속성(변수) 선택
X=df[['cement', 'slag','ash','water','superplastic','coarseagg','fineagg','age']]  #독립 변수 X

y=df['strength']                      #종속 변수 Y


'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''
#X.head()

# 설명 변수 데이터를 표준화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
#print(X)
#print(type(X))

# train data 와 test data로 구분(8:2 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

'''
[Step 5] 신경망 모형 - sklearn 사용
'''
##### from sklearn.neural_network
##### 1. 분류 : MLPClassifier
##### 2. 수치예측 : MLPRegressor

from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import GridSearchCV  # grid search 기능을 사용하기 위해서

param_grid = [
        {
            'activation' : ['identity', 'logistic', 'tanh', 'relu'],   # 활성화 함수의 종류 4가지
            'solver' : ['lbfgs', 'sgd', 'adam'],   # 옵티마이져(최적화)의 종류 3가지
            'hidden_layer_sizes': [
             (1,),(2,),(3,),(4,),(5,),(6,),(7,),(8,),(9,),(10,),(11,), (12,),(13,),(14,),(15,),(16,),(17,),(18,),(19,),(20,),(21,)
            ],
            'random_state' :[0,1,2,3,4,5,6,7,8,9,10]
            
        }
       ]


clf = GridSearchCV(MLPRegressor(), param_grid, cv=3, n_jobs = -1, verbose = 2 )


'''
[Step 6] train data를 가지고 모형 학습
'''
# train data를 가지고 모형 학습
clf.fit(X_train, y_train)   

print("Best parameters set found on development set:") 
print(clf.best_params_)  # 최고의 하이퍼 파라미터 조합을 출력해준다.
print('\n')


'''
[Step 7] test data를 가지고 y_hat을 예측
'''
# test data를 가지고 y_hat을 예측 (분류) 
y_hat = clf.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

'''
[Step 8] 모형 성능 평가 - 상관관계 계산
'''
import scipy.stats as stats
svm_report =  stats.pearsonr(y_test, y_hat)  # 상관계수 확인하는 코드
print(svm_report)  # 상관계수, p-value

train data 개수:  (824, 8)
test data 개수:  (206, 8)


Fitting 3 folds for each of 2772 candidates, totalling 8316 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  72 tasks      | elapsed:    5.8s
[Parallel(n_jobs=-1)]: Done 314 tasks      | elapsed:   27.1s
[Parallel(n_jobs=-1)]: Done 720 tasks      | elapsed:  1.1min
[Parallel(n_jobs=-1)]: Done 1286 tasks      | elapsed:  1.9min
[Parallel(n_jobs=-1)]: Done 2016 tasks      | elapsed:  3.0min
[Parallel(n_jobs=-1)]: Done 2906 tasks      | elapsed:  5.2min
[Parallel(n_jobs=-1)]: Done 3960 tasks      | elapsed:  8.5min
[Parallel(n_jobs=-1)]: Done 5174 tasks      | elapsed: 11.8min
[Parallel(n_jobs=-1)]: Done 6552 tasks      | elapsed: 16.1min
[Parallel(n_jobs=-1)]: Done 8090 tasks      | elapsed: 20.5min
[Parallel(n_jobs=-1)]: Done 8316 out of 8316 | elapsed: 21.2min finished


Best parameters set found on development set:
{'activation': 'relu', 'hidden_layer_sizes': (21,), 'random_state': 1, 'solver': 'lbfgs'}


[46.46770715 12.32803693 47.72641018 20.62907412 47.51553468 47.46104118
 11.80573824 57.42406157 23.0872247  44.06176696]
[47.4  10.35 49.99 14.7  47.82 51.06 13.29 52.82 22.5  47.74]


(0.9591328018424011, 8.227770552030871e-114)


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:470: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


Best parameters set found on development set:
{'activation': 'relu', 'hidden_layer_sizes': (21,), 'random_state': 1, 'solver': 'lbfgs'}   
상관계수, p-value : (0.9591328018424011, 8.227770552030871e-114)

문제42 grid search 를 이용하지 말고 보스톤 하우징 데이터의 수치를 예측하는 신경망 코드를 작성하시오 !

보스턴 데이터 설명:

    [01] CRIM 자치시(town) 별 1인당 범죄율 
    [02] ZN 25,000 평방피트를 초과하는 거주지역의 비율 
    [03] INDUS 비소매상업지역이 점유하고 있는 토지의 비율
    [04] CHAS 찰스강에 대한 더미변수(강의 경계에 위치한 경우는 1, 아니면 0) 
    [05] NOX 10ppm 당 농축 일산화질소 
    [06] RM 주택 1가구당 평균 방의 개수 
    [07] AGE 1940년 이전에 건축된 소유주택의 비율 
    [08] DIS 5개의 보스턴 직업센터까지의 접근성 지수 
    [09] RAD 방사형 도로까지의 접근성 지수 
    [10] TAX 10,000 달러 당 재산세율 
    [11] PTRATIO 자치시(town)별 학생/교사 비율 
    [12] B 1000(Bk-0.63)^2, 여기서 Bk는 자치시별 흑인의 비율을 말함. 
    [13] LSTAT 모집단의 하위계층의 비율(%) 
    [14] MEDV 본인 소유의 주택가격(중앙값) (단위: $1,000)

In [34]:
### 기본 라이브러리 불러오기
import pandas as pd

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = pd.read_csv('/content/drive/MyDrive/data/boston.csv')
df

Unnamed: 0     CRIM    ZN  INDUS  ...  PTRATIO       B  LSTAT  price
0             0  0.00632  18.0   2.31  ...     15.3  396.90   4.98   24.0
1             1  0.02731   0.0   7.07  ...     17.8  396.90   9.14   21.6
2             2  0.02729   0.0   7.07  ...     17.8  392.83   4.03   34.7
3             3  0.03237   0.0   2.18  ...     18.7  394.63   2.94   33.4
4             4  0.06905   0.0   2.18  ...     18.7  396.90   5.33   36.2
..          ...      ...   ...    ...  ...      ...     ...    ...    ...
501         501  0.06263   0.0  11.93  ...     21.0  391.99   9.67   22.4
502         502  0.04527   0.0  11.93  ...     21.0  396.90   9.08   20.6
503         503  0.06076   0.0  11.93  ...     21.0  396.90   5.64   23.9
504         504  0.10959   0.0  11.93  ...     21.0  393.45   6.48   22.0
505         505  0.04741   0.0  11.93  ...     21.0  396.90   7.88   11.9

[506 rows x 15 columns]

In [35]:
'''
[Step 2] 데이터 탐색/ 전처리
'''
# 결측치 확인
df.isnull().sum()

Unnamed: 0    0
CRIM          0
ZN            0
INDUS         0
CHAS          0
NOX           0
RM            0
AGE           0
DIS           0
RAD           0
TAX           0
PTRATIO       0
B             0
LSTAT         0
price         0
dtype: int64

In [36]:
df.head()

Unnamed: 0     CRIM    ZN  INDUS  CHAS  ...    TAX  PTRATIO       B  LSTAT  price
0           0  0.00632  18.0   2.31   0.0  ...  296.0     15.3  396.90   4.98   24.0
1           1  0.02731   0.0   7.07   0.0  ...  242.0     17.8  396.90   9.14   21.6
2           2  0.02729   0.0   7.07   0.0  ...  242.0     17.8  392.83   4.03   34.7
3           3  0.03237   0.0   2.18   0.0  ...  222.0     18.7  394.63   2.94   33.4
4           4  0.06905   0.0   2.18   0.0  ...  222.0     18.7  396.90   5.33   36.2

[5 rows x 15 columns]

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  506 non-null    int64  
 1   CRIM        506 non-null    float64
 2   ZN          506 non-null    float64
 3   INDUS       506 non-null    float64
 4   CHAS        506 non-null    float64
 5   NOX         506 non-null    float64
 6   RM          506 non-null    float64
 7   AGE         506 non-null    float64
 8   DIS         506 non-null    float64
 9   RAD         506 non-null    float64
 10  TAX         506 non-null    float64
 11  PTRATIO     506 non-null    float64
 12  B           506 non-null    float64
 13  LSTAT       506 non-null    float64
 14  price       506 non-null    float64
dtypes: float64(14), int64(1)
memory usage: 59.4 KB


In [39]:
df.columns

Index(['Unnamed: 0', 'CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS',
       'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'price'],
      dtype='object')

In [40]:
'''
[Step 3] 분석에 사용할 속성을 선택
'''
# 속성(변수) 선택
X=df[['CRIM','ZN','INDUS','CHAS','NOX','RM','AGE','DIS','RAD','TAX','PTRATIO','B','LSTAT']]  # 독립 변수 X

y=df['price']       # 종속 변수 Y

In [41]:
'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''
X.head()

# 설명 변수 데이터를 표준화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
print(X)
print('\n')
print(type(X))
print('\n')

[[-0.41978194  0.28482986 -1.2879095  ... -1.45900038  0.44105193
  -1.0755623 ]
 [-0.41733926 -0.48772236 -0.59338101 ... -0.30309415  0.44105193
  -0.49243937]
 [-0.41734159 -0.48772236 -0.59338101 ... -0.30309415  0.39642699
  -1.2087274 ]
 ...
 [-0.41344658 -0.48772236  0.11573841 ...  1.17646583  0.44105193
  -0.98304761]
 [-0.40776407 -0.48772236  0.11573841 ...  1.17646583  0.4032249
  -0.86530163]
 [-0.41500016 -0.48772236  0.11573841 ...  1.17646583  0.44105193
  -0.66905833]]


<class 'numpy.ndarray'>




In [42]:
# train data 와 test data로 구분(8:2 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

train data 개수:  (404, 13)
test data 개수:  (102, 13)




In [43]:
'''
[Step 5] 신경망 모형 - sklearn 사용
'''
##### from sklearn.neural_network
##### 1. 분류 : MLPClassifier
##### 2. 수치예측 : MLPRegressor

from sklearn.neural_network import MLPRegressor

# neural network 모델 적합
clf = MLPRegressor(random_state=0, max_iter=500)  # 학습을 500번 하겠다.

In [44]:
'''
[Step 6] train data를 가지고 모형 학습
'''
# train data를 가지고 모형 학습
clf.fit(X_train, y_train)   

MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(100,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=500,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=0, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [45]:
'''
[Step 7] test data를 가지고 y_hat을 예측
'''
# test data를 가지고 y_hat을 예측 (분류) 
y_hat = clf.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

[26.43408617 33.88557132 27.87271742 27.10147336 19.47663534 15.9936627
 40.44812842 16.32055055 21.13874284 41.52334783]
[28.4 31.1 23.5 26.6 19.6 14.3 50.  14.3 20.7 37.6]




In [46]:
'''
[Step 8] 모형 성능 평가 - 상관관계 계산
'''
import scipy.stats as stats
svm_report =  stats.pearsonr(y_test, y_hat)  # 상관계수 확인하는 코드
print(svm_report)   # 상관계수, p-value

(0.9139367239598873, 5.975509766599684e-41)


문제43. 보스톤 하우징 데이터의 가격 수치 예측하는 신경망 코드를
	    grid Search 를 이용해서 작성하고 상관계수 값을 올리세요

In [49]:
### 기본 라이브러리 불러오기
import pandas as pd

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = pd.read_csv('/content/drive/MyDrive/data/boston.csv')
#df

'''
[Step 2] 데이터 탐색/ 전처리
'''
# 결측치 확인
#df.isnull().sum()
#df.head()
#df.info()
#df.columns

'''
[Step 3] 분석에 사용할 속성을 선택
'''
# 속성(변수) 선택
X=df[['CRIM','ZN','INDUS','CHAS','NOX','RM','AGE','DIS','RAD','TAX','PTRATIO','B','LSTAT']]  # 독립 변수 X

y=df['price']       # 종속 변수 Y


'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''
#X.head()

# 설명 변수 데이터를 표준화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
#print(X)
#print(type(X))

# train data 와 test data로 구분(8:2 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

'''
[Step 5] 신경망 모형 - sklearn 사용
'''
##### from sklearn.neural_network
##### 1. 분류 : MLPClassifier
##### 2. 수치예측 : MLPRegressor

from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import GridSearchCV  # grid search 기능을 사용하기 위해서

param_grid = [
        {
            'activation' : ['identity', 'logistic', 'tanh', 'relu'],   # 활성화 함수의 종류 4가지
            'solver' : ['lbfgs', 'sgd', 'adam'],   # 옵티마이져(최적화)의 종류 3가지
            'hidden_layer_sizes': [
             (1,),(2,),(3,),(4,),(5,),(6,),(7,),(8,),(9,),(10,),(11,), (12,),(13,),(14,),(15,),(16,),(17,),(18,),(19,),(20,),(21,)
            ],
            'random_state' :[ 0,1,2,3,4,5,6,7,8,9,10]
            
        }
       ]


clf = GridSearchCV(MLPRegressor(), param_grid, cv=3, n_jobs = -1, verbose = 2 )

'''
[Step 6] train data를 가지고 모형 학습
'''
# train data를 가지고 모형 학습
clf.fit(X_train, y_train)   
print('\n')

print("Best parameters set found on development set:") 
print(clf.best_params_)  # 최고의 하이퍼 파라미터 조합을 출력해준다.
print('\n')

'''
[Step 7] test data를 가지고 y_hat을 예측
'''
# test data를 가지고 y_hat을 예측 (분류) 
y_hat = clf.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

'''
[Step 8] 모형 성능 평가 - 상관관계 계산
'''
import scipy.stats as stats
svm_report =  stats.pearsonr(y_test, y_hat)  # 상관계수 확인하는 코드
print(svm_report)   # 상관계수, p-value

train data 개수:  (404, 13)
test data 개수:  (102, 13)


Fitting 3 folds for each of 2772 candidates, totalling 8316 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  37 tasks      | elapsed:    3.4s
[Parallel(n_jobs=-1)]: Done 246 tasks      | elapsed:   16.5s
[Parallel(n_jobs=-1)]: Done 652 tasks      | elapsed:   42.2s
[Parallel(n_jobs=-1)]: Done 1218 tasks      | elapsed:  1.3min
[Parallel(n_jobs=-1)]: Done 1948 tasks      | elapsed:  2.1min
[Parallel(n_jobs=-1)]: Done 2838 tasks      | elapsed:  3.7min
[Parallel(n_jobs=-1)]: Done 3892 tasks      | elapsed:  5.9min
[Parallel(n_jobs=-1)]: Done 5106 tasks      | elapsed:  8.4min
[Parallel(n_jobs=-1)]: Done 6484 tasks      | elapsed: 11.4min
[Parallel(n_jobs=-1)]: Done 8022 tasks      | elapsed: 14.7min
[Parallel(n_jobs=-1)]: Done 8313 out of 8316 | elapsed: 15.4min remaining:    0.3s




Best parameters set found on development set:
{'activation': 'relu', 'hidden_layer_sizes': (10,), 'random_state': 4, 'solver': 'lbfgs'}


[28.64605051 32.20010547 26.46687749 25.6170062  18.33725428 15.85838117
 45.03909903 15.44360305 22.72423419 43.37239923]
[28.4 31.1 23.5 26.6 19.6 14.3 50.  14.3 20.7 37.6]


(0.9270169593988205, 2.179952578866812e-44)


[Parallel(n_jobs=-1)]: Done 8316 out of 8316 | elapsed: 15.4min finished
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:470: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


Best parameters set found on development set:
{'activation': 'relu', 'hidden_layer_sizes': (10,), 'random_state': 4, 'solver': 'lbfgs'}  
상관계수, p-value: (0.9270169593988205, 2.179952578866812e-44)

# **■ 서포트 벡터 머신(Support Vector Machine, SVM)**

서포트 벡터 머신은 기계학습 분야의 하나로 패턴인식, 자료분석을 위한
지도학습 모델이며 주로 분류와 회귀분석을 위해 사용한다.  
서포트 벡터 머신 알고리즘은 데이터를 분류하는 가장 큰 폭을 가진 경계를
찾는 알고리즘이다.  
비선형 분류를 하기 위해서 주어진 데이터를 고차원 특징 공간으로
사상하는 작업이 필요한데, 이를 효율적으로 하기 위해 커널 트릭을 사용하기도 한다.

예제. 서포트 벡터 머신 기본코드

In [50]:
### 기본 라이브러리 불러오기
import pandas as pd
import seaborn as sns

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = sns.load_dataset('titanic')

#  IPython 디스플레이 설정 - 출력할 열의 개수 한도 늘리기
pd.set_option('display.max_columns', 15)

In [51]:
'''
[Step 2] 데이터 탐색/ 전처리
'''

# NaN값이 많은 deck 열을 삭제, embarked와 내용이 겹치는 embark_town 열을 삭제
rdf = df.drop(['deck', 'embark_town'], axis=1)  

# age 열에 나이 데이터가 없는 모든 행을 삭제 - age 열(891개 중 177개의 NaN 값)
rdf = rdf.dropna(subset=['age'], how='any', axis=0)  

# embarked 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 치환하기
most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()   
rdf['embarked'].fillna(most_freq, inplace=True)

In [52]:
'''
[Step 3] 분석에 사용할 속성을 선택
'''

# 분석에 활용할 열(속성)을 선택 
ndf = rdf[['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'embarked']]

# 원핫인코딩 - 범주형 데이터를 모형이 인식할 수 있도록 숫자형으로 변환
onehot_sex = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf, onehot_sex], axis=1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix='town')
ndf = pd.concat([ndf, onehot_embarked], axis=1)

ndf.drop(['sex', 'embarked'], axis=1, inplace=True)

In [53]:
'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''

# 속성(변수) 선택
X=ndf[['pclass', 'age', 'sibsp', 'parch', 'female', 'male', 'town_C', 'town_Q', 'town_S']]  # 독립 변수 X
y=ndf['survived']      # 종속 변수 Y

# 설명 변수 데이터를 정규화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)

# train data 와 test data로 구분(7:3 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

train data 개수:  (499, 9)
test data 개수:  (215, 9)




In [54]:
'''
[Step 5] SVM 분류 모형 - sklearn 사용
'''

# sklearn 라이브러리에서 SVM 분류 모형 가져오기
from sklearn import svm

# 모형 객체 생성 (kernel='rbf' 적용)
svm_model = svm.SVC(kernel='rbf')

# train data를 가지고 모형 학습
svm_model.fit(X_train, y_train)   

# test data를 가지고 y_hat을 예측 (분류) 
y_hat = svm_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

[0 0 1 0 0 0 1 0 0 0]
[0 0 1 0 0 1 1 1 0 0]




In [55]:
# 모형 성능 평가 - Confusion Matrix 계산
from sklearn import metrics 
svm_matrix = metrics.confusion_matrix(y_test, y_hat)  
print(svm_matrix)
print('\n')

# 모형 성능 평가 - 평가지표 계산
svm_report = metrics.classification_report(y_test, y_hat)            
print(svm_report)

[[120   5]
 [ 35  55]]


              precision    recall  f1-score   support

           0       0.77      0.96      0.86       125
           1       0.92      0.61      0.73        90

    accuracy                           0.81       215
   macro avg       0.85      0.79      0.80       215
weighted avg       0.83      0.81      0.81       215



문제44. 지금까지의 정확도는 0.81 인데 아이와 여저 먼저라는 파생변수를 추가하면 정확도가 더 올라가는지 확인하시오 ! 

In [56]:
### 기본 라이브러리 불러오기
import pandas as pd
import seaborn as sns

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = sns.load_dataset('titanic')

#  IPython 디스플레이 설정 - 출력할 열의 개수 한도 늘리기
pd.set_option('display.max_columns', 15)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [57]:
# 여자와 아이에 대한 파생변수 추가
mask = ( df.age < 10 ) | ( df.sex=='female') 
mask.astype(int) # True 를 1 로 변경하고 False 를 0 으로 변경
df['women_child'] = mask.astype(int)

In [63]:
'''
[Step 2] 데이터 탐색/ 전처리
​'''

# NaN값이 많은 deck 열을 삭제, embarked와 내용이 겹치는 embark_town 열을 삭제
rdf = df.drop(['deck', 'embark_town'], axis=1)  

# age 열에 나이 데이터가 없는 모든 행을 삭제 - age 열(891개 중 177개의 NaN 값)
rdf = rdf.dropna(subset=['age'], how='any', axis=0)  

# embarked 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 치환하기
most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()   
rdf['embarked'].fillna(most_freq, inplace=True)
rdf

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male alive  alone  women_child  
0      man        True    no  False            0  
1    woman       False   yes  False            1  
2    woman       False   yes   True            1  
3    woman       False   yes  False            1  
4      man        True    no   True            0  
..     ...         ...   ...    ...          ...  
885  woman       False    no  False            1  
886    man        True    no   True            0  
887  woman       False   yes   True            1  
889    man        True   yes   True            0  
890    man        True    no   True            0  

[714 rows x 14 columns]

In [64]:
'''
[Step 3] 분석에 사용할 속성을 선택
'''

# 분석에 활용할 열(속성)을 선택 
ndf = rdf[['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'embarked','women_child']]

# 원핫인코딩 - 범주형 데이터를 모형이 인식할 수 있도록 숫자형으로 변환
onehot_sex = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf, onehot_sex], axis=1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix='town')
ndf = pd.concat([ndf, onehot_embarked], axis=1)

ndf.drop(['sex', 'embarked'], axis=1, inplace=True)
ndf

survived  pclass   age  sibsp  parch  women_child  female  male  town_C  \
0           0       3  22.0      1      0            0       0     1       0   
1           1       1  38.0      1      0            1       1     0       1   
2           1       3  26.0      0      0            1       1     0       0   
3           1       1  35.0      1      0            1       1     0       0   
4           0       3  35.0      0      0            0       0     1       0   
..        ...     ...   ...    ...    ...          ...     ...   ...     ...   
885         0       3  39.0      0      5            1       1     0       0   
886         0       2  27.0      0      0            0       0     1       0   
887         1       1  19.0      0      0            1       1     0       0   
889         1       1  26.0      0      0            0       0     1       1   
890         0       3  32.0      0      0            0       0     1       0   

     town_Q  town_S  
0         0       1  
1         0       0  
2         0       1  
3         0       1  
4         0       1  
..      ...     ...  
885       1       0  
886       0       1  
887       0       1  
889       0       0  
890       1       0  

[714 rows x 11 columns]

In [65]:
'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''

# 속성(변수) 선택
X=ndf[['pclass', 'age', 'sibsp', 'parch', 'female', 'male', 
       'town_C', 'town_Q', 'town_S','women_child']]  #독립 변수 X
y=ndf['survived']                      #종속 변수 Y

# 설명 변수 데이터를 정규화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)

# train data 와 test data로 구분(7:3 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

train data 개수:  (499, 10)
test data 개수:  (215, 10)




In [66]:
'''
[Step 5] SVM 분류 모형 - sklearn 사용
'''

# sklearn 라이브러리에서 SVM 분류 모형 가져오기
from sklearn import svm

# 모형 객체 생성 (kernel='rbf' 적용)
svm_model = svm.SVC(kernel='rbf')

# train data를 가지고 모형 학습
svm_model.fit(X_train, y_train)   

# test data를 가지고 y_hat을 예측 (분류) 
y_hat = svm_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

[0 0 1 0 0 1 1 0 0 0]
[0 0 1 0 0 1 1 1 0 0]




In [67]:
# 모형 성능 평가 - Confusion Matrix 계산
from sklearn import metrics 
svm_matrix = metrics.confusion_matrix(y_test, y_hat)  
print(svm_matrix)
print('\n')

# 모형 성능 평가 - 평가지표 계산
svm_report = metrics.classification_report(y_test, y_hat)            
print(svm_report)

[[120   5]
 [ 31  59]]


              precision    recall  f1-score   support

           0       0.79      0.96      0.87       125
           1       0.92      0.66      0.77        90

    accuracy                           0.83       215
   macro avg       0.86      0.81      0.82       215
weighted avg       0.85      0.83      0.83       215



문제45. grid search 를 이용해서 최적의 하이퍼 파라미터를 알아내시오 !

서포트 벡터 머신의 하이퍼 파라미터 ?  C 와 gamma 와 kernel

In [68]:
### 기본 라이브러리 불러오기
import pandas as pd
import seaborn as sns

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = sns.load_dataset('titanic')

#  IPython 디스플레이 설정 - 출력할 열의 개수 한도 늘리기
pd.set_option('display.max_columns', 15)
#df

# 여자와 아이에 대한 파생변수 추가
mask = ( df.age < 10 ) | ( df.sex=='female') 
mask.astype(int) # True 를 1 로 변경하고 False 를 0 으로 변경
df['women_child'] = mask.astype(int)

'''
[Step 2] 데이터 탐색/ 전처리
'''

# NaN값이 많은 deck 열을 삭제, embarked와 내용이 겹치는 embark_town 열을 삭제
rdf = df.drop(['deck', 'embark_town'], axis=1)  

# age 열에 나이 데이터가 없는 모든 행을 삭제 - age 열(891개 중 177개의 NaN 값)
rdf = rdf.dropna(subset=['age'], how='any', axis=0)  

# embarked 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 치환하기
most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()   
rdf['embarked'].fillna(most_freq, inplace=True)
#rdf

'''
[Step 3] 분석에 사용할 속성을 선택
'''

# 분석에 활용할 열(속성)을 선택 
ndf = rdf[['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'embarked','women_child']]

# 원핫인코딩 - 범주형 데이터를 모형이 인식할 수 있도록 숫자형으로 변환
onehot_sex = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf, onehot_sex], axis=1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix='town')
ndf = pd.concat([ndf, onehot_embarked], axis=1)

ndf.drop(['sex', 'embarked'], axis=1, inplace=True)
#ndf

'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''

# 속성(변수) 선택
X=ndf[['pclass', 'age', 'sibsp', 'parch', 'female', 'male', 
       'town_C', 'town_Q', 'town_S','women_child']]  #독립 변수 X
y=ndf['survived']                      #종속 변수 Y

# 설명 변수 데이터를 정규화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)

# train data 와 test data로 구분(7:3 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

'''
[Step 5] SVM 분류 모형 - sklearn 사용
'''

# sklearn 라이브러리에서 SVM 분류 모형 가져오기
from sklearn import svm
from sklearn.model_selection import GridSearchCV 
  
# defining parameter range 
param_grid = {'C': [0.1, 1, 10, 100, 1000],  
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001], 
              'kernel': ['rbf','poly', 'sigmoid','linear']}  
  
grid = GridSearchCV( svm.SVC(), param_grid, refit = True,cv=3, n_jobs = -1, verbose = 2 )

# fitting the model for grid search 
grid.fit(X_train, y_train) 
print('\n')

# test data를 가지고 y_hat을 예측 (분류) 
y_hat = grid.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

# 모형 성능 평가 - Confusion Matrix 계산
from sklearn import metrics 
svm_matrix = metrics.confusion_matrix(y_test, y_hat)  
print(svm_matrix)
print('\n')

# 모형 성능 평가 - 평가지표 계산
svm_report = metrics.classification_report(y_test, y_hat)            
print(svm_report)

train data 개수:  (499, 10)
test data 개수:  (215, 10)


Fitting 3 folds for each of 100 candidates, totalling 300 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done 186 tasks      | elapsed:    6.9s




[0 0 1 0 0 1 1 0 0 0]
[0 0 1 0 0 1 1 1 0 0]


[[120   5]
 [ 31  59]]


              precision    recall  f1-score   support

           0       0.79      0.96      0.87       125
           1       0.92      0.66      0.77        90

    accuracy                           0.83       215
   macro avg       0.86      0.81      0.82       215
weighted avg       0.85      0.83      0.83       215



[Parallel(n_jobs=-1)]: Done 300 out of 300 | elapsed:  4.7min finished


문제46. 서포트 벡터 머신으로 최고로 올릴 수 있는 타이타닉 데이터의 정확도는 0.83 이었습니다. 위에서 했던 gridsearch 를 이용한 신경망 코드를 이용해서 수치예측이 아닌 분류로 머신러닝 함수를 변경해서 훈련시키고 정확도를 확인하시오 !

 1. 분류 : MLPClassifier
 2. 수치예측 : MLPRegressor
from sklearn.neural_network import MLPClassifier  ----->  분류
from sklearn.neural_network import MLPRegressor  ----->  수치예측


In [ ]:
#힌트코드:
'''
[Step 5] 신경망 분류 모형 - sklearn 사용
'''

from sklearn.neural_network  import  MLPClassifier
from sklearn.model_selection import GridSearchCV

param_grid = [
        {
            'activation' : ['identity', 'logistic', 'tanh', 'relu'],
            'solver' : ['lbfgs', 'sgd', 'adam'],
            'hidden_layer_sizes': [
             (1,),(2,),(3,),(4,),(5,),(6,),(7,),(8,),(9,),(10,),(11,), (12,),(13,),(14,),(15,),(16,),(17,),(18,),(19,),(20,),(21,)
             ],
            'random_state' :[ 0,1,2,3,4,5,6,7,8,9,10]
        }
       ]
 
grid = GridSearchCV(MLPClassifier(), param_grid, cv=3, n_jobs = -1, verbose = 2 )

In [70]:
### 기본 라이브러리 불러오기
import pandas as pd
import seaborn as sns

'''
[Step 1] 데이터 준비/ 기본 설정
'''
# load_dataset 함수를 사용하여 데이터프레임으로 변환
df = sns.load_dataset('titanic')

#  IPython 디스플레이 설정 - 출력할 열의 개수 한도 늘리기
pd.set_option('display.max_columns', 15)
#df

# 여자와 아이에 대한 파생변수 추가
mask = ( df.age < 10 ) | ( df.sex=='female') 
mask.astype(int) # True 를 1 로 변경하고 False 를 0 으로 변경
df['women_child'] = mask.astype(int)

'''
[Step 2] 데이터 탐색/ 전처리
'''

# NaN값이 많은 deck 열을 삭제, embarked와 내용이 겹치는 embark_town 열을 삭제
rdf = df.drop(['deck', 'embark_town'], axis=1)  

# age 열에 나이 데이터가 없는 모든 행을 삭제 - age 열(891개 중 177개의 NaN 값)
rdf = rdf.dropna(subset=['age'], how='any', axis=0)  

# embarked 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 치환하기
most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()   
rdf['embarked'].fillna(most_freq, inplace=True)
#rdf

'''
[Step 3] 분석에 사용할 속성을 선택
'''

# 분석에 활용할 열(속성)을 선택 
ndf = rdf[['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'embarked','women_child']]

# 원핫인코딩 - 범주형 데이터를 모형이 인식할 수 있도록 숫자형으로 변환
onehot_sex = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf, onehot_sex], axis=1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix='town')
ndf = pd.concat([ndf, onehot_embarked], axis=1)

ndf.drop(['sex', 'embarked'], axis=1, inplace=True)
#ndf

'''
[Step 4] 데이터셋 구분 - 훈련용(train data)/ 검증용(test data)
'''

# 속성(변수) 선택
X=ndf[['pclass', 'age', 'sibsp', 'parch', 'female', 'male', 
       'town_C', 'town_Q', 'town_S','women_child']]  #독립 변수 X
y=ndf['survived']                      #종속 변수 Y

# 설명 변수 데이터를 정규화(normalization)
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)

# train data 와 test data로 구분(7:3 비율)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=10) 

print('train data 개수: ', X_train.shape)
print('test data 개수: ', X_test.shape)
print('\n')

'''
[Step 5] 신경망 분류 모형 - sklearn 사용
'''

from sklearn.neural_network  import  MLPClassifier
from sklearn.model_selection import GridSearchCV

param_grid = [
        {
            'activation' : ['identity', 'logistic', 'tanh', 'relu'],
            'solver' : ['lbfgs', 'sgd', 'adam'],
            'hidden_layer_sizes': [
             (1,),(2,),(3,),(4,),(5,),(6,),(7,),(8,),(9,),(10,),(11,), (12,),(13,),(14,),(15,),(16,),(17,),(18,),(19,),(20,),(21,)
             ],
            'random_state' :[ 0,1,2,3,4,5,6,7,8,9,10]
        }
       ]
 
grid = GridSearchCV(MLPClassifier(), param_grid, cv=3, n_jobs = -1, verbose = 2 )

# fitting the model for grid search 
grid.fit(X_train, y_train) 
print('\n')

print("Best parameters set found on development set:")
print(grid.best_params_) 
print('\n')

# test data를 가지고 y_hat을 예측 (분류) 
y_hat = grid.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

# 모형 성능 평가 - Confusion Matrix 계산
from sklearn import metrics 
svm_matrix = metrics.confusion_matrix(y_test, y_hat)  
print(svm_matrix)
print('\n')

# 모형 성능 평가 - 평가지표 계산
svm_report = metrics.classification_report(y_test, y_hat)            
print(svm_report)

train data 개수:  (499, 10)
test data 개수:  (215, 10)


Fitting 3 folds for each of 2772 candidates, totalling 8316 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  37 tasks      | elapsed:    4.2s
[Parallel(n_jobs=-1)]: Done 158 tasks      | elapsed:   14.6s
[Parallel(n_jobs=-1)]: Done 361 tasks      | elapsed:   32.9s
[Parallel(n_jobs=-1)]: Done 644 tasks      | elapsed:   58.3s
[Parallel(n_jobs=-1)]: Done 1009 tasks      | elapsed:  1.5min
[Parallel(n_jobs=-1)]: Done 1454 tasks      | elapsed:  2.2min
[Parallel(n_jobs=-1)]: Done 1981 tasks      | elapsed:  3.0min
[Parallel(n_jobs=-1)]: Done 3072 tasks      | elapsed:  5.1min
[Parallel(n_jobs=-1)]: Done 4450 tasks      | elapsed:  8.2min
[Parallel(n_jobs=-1)]: Done 5988 tasks      | elapsed: 11.7min
[Parallel(n_jobs=-1)]: Done 7690 tasks      | elapsed: 15.6min
[Parallel(n_jobs=-1)]: Done 8316 out of 8316 | elapsed: 17.1min finished




Best parameters set found on development set:
{'activation': 'relu', 'hidden_layer_sizes': (17,), 'random_state': 8, 'solver': 'adam'}


[0 0 1 0 0 1 1 0 0 0]
[0 0 1 0 0 1 1 1 0 0]


[[120   5]
 [ 26  64]]


              precision    recall  f1-score   support

           0       0.82      0.96      0.89       125
           1       0.93      0.71      0.81        90

    accuracy                           0.86       215
   macro avg       0.87      0.84      0.85       215
weighted avg       0.87      0.86      0.85       215



/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


# **■ 연관규칙(Association Rules)**

간단한 성능 측정치(지지도, 신뢰도, 향상도) 를 이용해서 거대한 데이터에서
데이터간의 연관성을 찾는 알고리즘

예제:

In [71]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder  # 연관성 분석을 위한 데이터 전처리
from mlxtend.frequent_patterns import apriori # 연관성 분석을 위한 아프리오리 알고리즘 함수 

dataset=[['사과','치즈','생수'],
['생수','호두','치즈','고등어'],
['수박','사과','생수'],
['생수','호두','치즈','옥수수']]

te = TransactionEncoder()
te_ary = te.fit(dataset).transform(dataset)
print(te.columns_)
te_ary

df = pd.DataFrame(te_ary, columns=te.columns_)  # 보기 좋게 데이터프레임으로 변경
df

frequent_itemsets = apriori(df, min_support=0.5, use_colnames=True)  # 지지도 0.5 이상
print(frequent_itemsets )   # 사람들이 많이 산 물건들만 출력

from mlxtend.frequent_patterns import association_rules # 연관성을 찾는 함수 import
print( association_rules(frequent_itemsets, metric="confidence", min_threshold=0.3) ) 
# 신뢰도를 기준으로 물품들간의 연관성을 출력하는데 임계치를 0.3 이상인 것만 출력

['고등어', '사과', '생수', '수박', '옥수수', '치즈', '호두']
   support      itemsets
0     0.50          (사과)
1     1.00          (생수)
2     0.75          (치즈)
3     0.50          (호두)
4     0.50      (생수, 사과)
5     0.75      (생수, 치즈)
6     0.50      (생수, 호두)
7     0.50      (호두, 치즈)
8     0.50  (생수, 호두, 치즈)
   antecedents consequents  antecedent support  consequent support  support  \
0         (생수)        (사과)                1.00                0.50     0.50   
1         (사과)        (생수)                0.50                1.00     0.50   
2         (생수)        (치즈)                1.00                0.75     0.75   
3         (치즈)        (생수)                0.75                1.00     0.75   
4         (생수)        (호두)                1.00                0.50     0.50   
5         (호두)        (생수)                0.50                1.00     0.50   
6         (호두)        (치즈)                0.50                0.75     0.50   
7         (치즈)        (호두)                0.75                0.50     0.

결론:
생수와 사과는 같은 곳에 진열을 해야한다.

문제47. 아래의 데이터에서 연관성을 도출하시오 !

    dataset=[['빵','우유'],
            ['맥주','빵','기저귀','계란'],
           ['맥주','콜라','기저귀','우유'],
           ['콜라','빵','기저귀','우유']]

In [72]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori 

dataset=[['빵','우유'],
            ['맥주','빵','기저귀','계란'],
           ['맥주','콜라','기저귀','우유'],
           ['콜라','빵','기저귀','우유']]

te = TransactionEncoder()
te_ary = te.fit(dataset).transform(dataset)
print(te.columns_)
te_ary

df = pd.DataFrame(te_ary, columns=te.columns_)  
df

frequent_itemsets = apriori(df, min_support=0.5, use_colnames=True)  
print(frequent_itemsets ) 

from mlxtend.frequent_patterns import association_rules 
print( association_rules(frequent_itemsets, metric="confidence", min_threshold=0.3) ) 

['계란', '기저귀', '맥주', '빵', '우유', '콜라']
    support       itemsets
0      0.75          (기저귀)
1      0.50           (맥주)
2      0.75            (빵)
3      0.75           (우유)
4      0.50           (콜라)
5      0.50      (맥주, 기저귀)
6      0.50       (빵, 기저귀)
7      0.50      (우유, 기저귀)
8      0.50      (기저귀, 콜라)
9      0.50        (빵, 우유)
10     0.50       (우유, 콜라)
11     0.50  (우유, 기저귀, 콜라)
   antecedents consequents  antecedent support  consequent support  support  \
0         (맥주)       (기저귀)                0.50                0.75      0.5   
1        (기저귀)        (맥주)                0.75                0.50      0.5   
2          (빵)       (기저귀)                0.75                0.75      0.5   
3        (기저귀)         (빵)                0.75                0.75      0.5   
4         (우유)       (기저귀)                0.75                0.75      0.5   
5        (기저귀)        (우유)                0.75                0.75      0.5   
6        (기저귀)        (콜라)                0.75             

문제48. 보습학원 데이터로 연관분석을 
	   해보겠습니다. 보습학원이 있는 건물에는 어떤 업종이 많이 있는가?
	 

In [74]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
import numpy as np

# CSV 파일을 데이터프레임으로 변환
dataset = pd.read_csv('/content/drive/MyDrive/data/building3.csv', encoding='cp949', header=None)
dataset = dataset[ [1,2,3,4,5,6,7,8,9,10,11,12]]
dataset = dataset.fillna(0)
dataset

1    2    3    4    5    6    7    8    9    10   11   12
0   1.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0
1   0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  1.0  0.0
2   0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0
3   0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0
4   1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0
5   0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
6   0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0
7   1.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
8   0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0
9   0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0
10  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0
11  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0
12  1.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
13  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0
14  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0
15  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0
16  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0
17  1.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
18  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0
19  1.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

In [75]:
import  numpy  as  np

d = np.array(dataset)
d2 = d.astype('bool')
d2

array([[ True,  True, False,  True, False, False, False, False, False,
        False, False,  True],
       [False, False, False, False,  True,  True,  True, False, False,
        False,  True, False],
       [False, False,  True, False, False, False, False,  True, False,
         True, False, False],
       [False, False,  True, False, False, False, False,  True, False,
         True, False, False],
       [ True, False, False, False,  True,  True, False, False, False,
        False,  True,  True],
       [False, False,  True,  True, False, False, False, False, False,
        False, False, False],
       [False, False, False, False, False,  True, False, False, False,
        False, False,  True],
       [ True,  True, False,  True, False, False, False, False, False,
        False, False, False],
       [False, False, False, False, False, False,  True, False,  True,
        False, False, False],
       [False, False, False, False,  True,  True, False, False, False,
        False,  True

In [76]:
df = pd.DataFrame(d2, columns=['hospital', 'pharmacy', 'cafe', 'mobile', 'restaurant',
       'familyrest', 'billiardroom', 'academy', 'supermarket', 'bank',
       'convenience', 'cosmetics']) #위에서 나온걸 보기 좋게 데이터프레임으로 변경
df

hospital  pharmacy   cafe  mobile  restaurant  familyrest  billiardroom  \
0       True      True  False    True       False       False         False   
1      False     False  False   False        True        True          True   
2      False     False   True   False       False       False         False   
3      False     False   True   False       False       False         False   
4       True     False  False   False        True        True         False   
5      False     False   True    True       False       False         False   
6      False     False  False   False       False        True         False   
7       True      True  False    True       False       False         False   
8      False     False  False   False       False       False          True   
9      False     False  False   False        True        True         False   
10     False     False   True   False       False       False         False   
11     False     False  False   False        True        True          True   
12      True      True  False    True       False       False         False   
13     False     False  False   False        True        True         False   
14     False     False  False   False        True        True         False   
15     False     False   True   False       False       False         False   
16     False     False  False   False        True        True          True   
17      True      True  False    True       False       False         False   
18     False     False  False   False        True        True          True   
19      True      True  False    True       False       False         False   

    academy  supermarket   bank  convenience  cosmetics  
0     False        False  False        False       True  
1     False        False  False         True      False  
2      True        False   True        False      False  
3      True        False   True        False      False  
4     False        False  False         True       True  
5     False        False  False        False      False  
6     False        False  False        False       True  
7     False        False  False        False      False  
8     False         True  False        False      False  
9     False        False  False         True       True  
10     True        False   True        False      False  
11    False        False  False        False      False  
12    False        False  False        False      False  
13    False        False  False         True       True  
14    False        False  False         True       True  
15     True        False   True        False      False  
16    False        False  False        False      False  
17    False        False  False        False      False  
18    False        False  False        False      False  
19    False        False  False        False      False

In [77]:
frequent_itemsets = apriori(df, min_support=0.01, use_colnames=True)
print(frequent_itemsets)

    support                                           itemsets
0      0.30                                         (hospital)
1      0.25                                         (pharmacy)
2      0.25                                             (cafe)
3      0.30                                           (mobile)
4      0.40                                       (restaurant)
..      ...                                                ...
56     0.05     (restaurant, convenience, hospital, cosmetics)
57     0.05     (familyrest, convenience, hospital, cosmetics)
58     0.05  (familyrest, billiardroom, convenience, restau...
59     0.20   (familyrest, restaurant, convenience, cosmetics)
60     0.05  (familyrest, convenience, cosmetics, hospital,...

[61 rows x 2 columns]


In [78]:
from mlxtend.frequent_patterns import association_rules
rules1 = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.1) 
rules1

antecedents                                       consequents  \
0       (pharmacy)                                        (hospital)   
1       (hospital)                                        (pharmacy)   
2         (mobile)                                        (hospital)   
3       (hospital)                                          (mobile)   
4       (hospital)                                      (restaurant)   
..             ...                                               ...   
277   (familyrest)    (restaurant, convenience, hospital, cosmetics)   
278  (convenience)     (familyrest, restaurant, hospital, cosmetics)   
279    (cosmetics)   (familyrest, convenience, hospital, restaurant)   
280     (hospital)  (familyrest, restaurant, convenience, cosmetics)   
281   (restaurant)    (familyrest, convenience, hospital, cosmetics)   

     antecedent support  consequent support  support  confidence      lift  \
0                  0.25                0.30     0.25    1.000000  3.333333   
1                  0.30                0.25     0.25    0.833333  3.333333   
2                  0.30                0.30     0.25    0.833333  2.777778   
3                  0.30                0.30     0.25    0.833333  2.777778   
4                  0.30                0.40     0.05    0.166667  0.416667   
..                  ...                 ...      ...         ...       ...   
277                0.45                0.05     0.05    0.111111  2.222222   
278                0.25                0.05     0.05    0.200000  4.000000   
279                0.30                0.05     0.05    0.166667  3.333333   
280                0.30                0.20     0.05    0.166667  0.833333   
281                0.40                0.05     0.05    0.125000  2.500000   

     leverage  conviction  
0      0.1750         inf  
1      0.1750    4.500000  
2      0.1600    4.200000  
3      0.1600    4.200000  
4     -0.0700    0.720000  
..        ...         ...  
277    0.0275    1.068750  
278    0.0375    1.187500  
279    0.0350    1.140000  
280   -0.0100    0.960000  
281    0.0300    1.085714  

[282 rows x 9 columns]

In [79]:
rules1[rules1.antecedents.apply(str).str.contains("academy")].sort_values('lift', ascending=False)
# 참고 : https://ichi.pro/ko/python-ui-sijang-baguni-bunseog-e-daehan-jaseubseo-80804556335911

antecedents   consequents  antecedent support  consequent support  \
43         (academy)        (bank)                 0.2                0.20   
108  (cafe, academy)        (bank)                 0.2                0.20   
110        (academy)  (bank, cafe)                 0.2                0.20   
18         (academy)        (cafe)                 0.2                0.25   
106  (bank, academy)        (cafe)                 0.2                0.25   

     support  confidence  lift  leverage  conviction  
43       0.2         1.0   5.0      0.16         inf  
108      0.2         1.0   5.0      0.16         inf  
110      0.2         1.0   5.0      0.16         inf  
18       0.2         1.0   4.0      0.15         inf  
106      0.2         1.0   4.0      0.15         inf